exercises/non-retryable-errors-types/README.md

```diff
@@ -0,0 +1,70 @@
+
+## Part C: Add Heartbeats
+
+In this part of the exercise, we will add heartbeating to our `pollDeliveryDriver` Activity.
+
+1. Edit `activities.ts`. We have added a `pollDeliveryDriver` Activity. This Activity polls an external service for delivery drivers. If that service returns a status code of 500s or 403, we don't want to retry polling this service. Within this Activity, within the `if` statement that checks the status code, throw a new `Application Failure` with a message that lets the user know that there is an invalid server error. Set this Application Failure's `nonRetryable` key to true.
+2. Now, let's add heartbeating. Import `heartbeat` and `activityInfo` from `@temporalio/activity`.
+3. In the `pollExternalDeliveryDriver` Activity, notice that we have a `startingPoint` variable. This variable is set to the resuming point that the heartbeat last left off of, or 1, if the heartbeating has not began.
+4. Add your entire `try/catch` block into a `for loop`. When initiating the loop, it should initiate at `let progress = startingPoint` and the progress should increment by one after each iteration of the loop. The loop should iterate up to ten times, one by one. This loop will simulate multiple attempts to poll an external service (e.g., DoorDash, UberEats) to find an available delivery driver.
+5. Call `heartbeat()` within the `for loop` so it invokes in each iteration of the loop. The `heartbeat` function should take in `progress`.
+6. Add a break statement after 'log.info(`External delivery driver assigned from: ${content.service}`)', so that we don't keep polling if the response is successful.
+7. Save your file. 
+
+## Part D: Add a Heartbeat Timeout
+
+In this part of the exercise, we will add a Heartbeat Timeout to your Activities. If a heartbeat Timeout is not set, Temporal doesn't track the Heartbeats sent by the Activity. 
+
+1. Edit `workflows.ts`.
+2. Below the `startToCloseTimeout`, add a `heartbeatTimeout` and set it to 30 seconds. This sets the maximum time between Activity Heartbeats. If an Activity times out (e.g., due to a missed Heartbeat), the next attempt can use this payload to continue from where it left off.
+3. Save your file.
+
+## Part E: Run the Workflow
+
+Next, let's run the Workflow.
+
+1. In one terminal, start the service that will poll for external delivery drivers by running `npm run service`.
+2. In another terminal, run the Worker by running `npm run start`.
+3. In another terminal, start the Workflow by running `npm run workflow`.
+
+Before your Workflow is completed, if you click on the Workflow ID in your Web UI, you can view the most recent Heartbeat message for a pending Activity under 'Pending Activities'. The Heartbeat message is not visible in the Web UI for an Activity Execution that has closed.
+
```

Comment:
Up to you, but I added an optional Fail the Workflow in Java. We don't necessarily need this for replay, but we should add it back in prior to LMS release.